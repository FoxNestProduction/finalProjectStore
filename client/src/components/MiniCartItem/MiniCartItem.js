import React, { memo, useCallback } from 'react';
import PropTypes from 'prop-types';
import {
  Box,
  Card,
  CardContent,
  CardMedia,
  Typography,
  ButtonGroup,
  Button,

} from '@mui/material';
import RemoveRoundedIcon from '@mui/icons-material/RemoveRounded';
import AddRounded from '@mui/icons-material/AddRounded';
import { useDispatch, useSelector } from 'react-redux';
import {
  cartProductImg,
  MiniCartItemContainer,
  buttonStyles,
  quantityStyle,
  textAndQuantityCardContent,
  cardTextContent,
  productName,
  currentPriceStyles,
  buttonsWrapper,
  roundedIcons,
} from './styles';
import { deleteFromCart, addOneMore } from '../../redux/slices/cartSlice';
import { totalSumFromCartProduct } from '../Cart/cartFunctions';

const MiniCartItem = ({ _id, name, cartQuantity, currentPrice, imageUrl }) => {
  const dispatch = useDispatch();
  const cartProducts = useSelector((state) => state.cart.cart.products);
  const index = cartProducts.findIndex(({ product }) => product._id === _id);
  const relevantPrice = totalSumFromCartProduct(currentPrice, cartQuantity);
  const handleDeleteOne = useCallback(() => {
    if (index !== -1) {
      const foundObject = cartProducts[index];
      dispatch(deleteFromCart(foundObject));
    }
  }, [cartProducts, dispatch, index]);
  const handleAddOne = useCallback(() => {
    if (index !== -1) {
      const foundObject = cartProducts[index];
      dispatch(addOneMore(foundObject));
    }
  }, [cartProducts, dispatch, index]);
  return (
    <Card
      sx={MiniCartItemContainer}
    >
      <Box>
        <CardMedia
          component="img"
          sx={cartProductImg}
          image={imageUrl}
          alt={name}
        />
      </Box>
      <Box
        sx={textAndQuantityCardContent}
      >
        <Box
          sx={{
            alignSelf: 'center',
          }}
        >
          <CardContent
            sx={cardTextContent}
          >
            <Typography
              sx={productName}
            >
              {name}
            </Typography>
            <Typography
              sx={currentPriceStyles}
            >
              $
              {relevantPrice}
            </Typography>
          </CardContent>
        </Box>
        <Box
          sx={buttonsWrapper}
        >
          <ButtonGroup
            variant="outlined"
            aria-label="small button group"
          >
            <Button
              onClick={handleDeleteOne}
              sx={buttonStyles}
            >
              <RemoveRoundedIcon
                sx={roundedIcons}
              />
            </Button>
            <Typography
              sx={quantityStyle}
            >
              {cartQuantity}
            </Typography>
            <Button
              onClick={handleAddOne}
              sx={buttonStyles}
            >
              <AddRounded
                sx={roundedIcons}
              />
            </Button>
          </ButtonGroup>

        </Box>
      </Box>
    </Card>
  );
};

MiniCartItem.propTypes = {
  _id: PropTypes.string,
  name: PropTypes.string.isRequired,
  cartQuantity: PropTypes.number,
  currentPrice: PropTypes.number.isRequired,
  imageUrl: PropTypes.string,
};

MiniCartItem.defaultProps = {
  _id: '',
  cartQuantity: 1,
  imageUrl: './img/pasta.png',
};

export default memo(MiniCartItem);
