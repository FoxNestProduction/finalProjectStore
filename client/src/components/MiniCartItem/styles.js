export const MiniCartItemContainer = {
  display: {
    mobile: 'none',
    tablet: 'flex',
  },
};

export const cartProductImgWrapper = {

};

export const cartProductImg = {
  width: {
    tablet: '40px',
  },
  height: {
    tablet: '40px',
  },
};
