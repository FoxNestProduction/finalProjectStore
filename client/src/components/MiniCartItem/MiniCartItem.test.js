import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, fireEvent, screen } from '@testing-library/react';
import MiniCartItem from './MiniCartItem';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

const mockDispatch = jest.spyOn(require('react-redux'), 'useDispatch');

describe('MiniCartItem Component', () => {

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should renders MiniCartItem component', () => {
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce([]);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(true);

    const { asFragment } = render(
      <MemoryRouter>
          <MiniCartItem
            name="Product"
            currentPrice={10.99}
            imageUrl="image.jpg"
            cartQuantity={1}
            _id="123"
          />
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('should delete product product when click dell', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce([]);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(true);

    render(
      <MemoryRouter>
          <MiniCartItem
            name="Product"
            currentPrice={10.99}
            imageUrl="image.jpg"
            cartQuantity={1}
            _id="123"
          />
      </MemoryRouter>,
    );

    const deleteButtons = screen.getAllByTestId('RemoveRoundedIcon');
    fireEvent.click(deleteButtons[0]);
    expect(dispatch).toHaveBeenCalled();
  });

  test('should added one product when click plus', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce([]);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(true);

    render(
      <MemoryRouter>
          <MiniCartItem
            name="Product"
            currentPrice={10.99}
            imageUrl="image.jpg"
            cartQuantity={1}
            _id="123"
          />
      </MemoryRouter>,
    );


    const addButtons = screen.getAllByTestId('AddRoundedIcon');
    fireEvent.click(addButtons[0]);
    expect(dispatch).toHaveBeenCalled();
  });
});