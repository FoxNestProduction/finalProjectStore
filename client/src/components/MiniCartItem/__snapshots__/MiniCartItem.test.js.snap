// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniCartItem Component should renders MiniCartItem component 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-82itc0-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiBox-root css-0"
    >
      <img
        alt="Product"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-12fewpf-MuiCardMedia-root"
        src="image.jpg"
      />
    </div>
    <div
      class="MuiBox-root css-1oeopgt"
    >
      <div
        class="MuiBox-root css-hoe9xz"
      >
        <div
          class="MuiCardContent-root css-1tallcb-MuiCardContent-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-qa37qg-MuiTypography-root"
          >
            Product
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-mlpg4e-MuiTypography-root"
          >
            $10.99
          </p>
        </div>
      </div>
      <div
        class="MuiBox-root css-1txf5x2"
      >
        <div
          aria-label="small button group"
          class="MuiButtonGroup-root MuiButtonGroup-outlined css-edsu1d-MuiButtonGroup-root"
          role="group"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-firstButton css-hu2o6-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1n2fl28-MuiSvgIcon-root"
              data-testid="RemoveRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 13H6c-.55 0-1-.45-1-1s.45-1 1-1h12c.55 0 1 .45 1 1s-.45 1-1 1z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1579pud-MuiTypography-root"
          >
            1
          </p>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-lastButton css-hu2o6-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1n2fl28-MuiSvgIcon-root"
              data-testid="AddRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 13h-5v5c0 .55-.45 1-1 1s-1-.45-1-1v-5H6c-.55 0-1-.45-1-1s.45-1 1-1h5V6c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
